zcash_client_sqlite/src/wallet/transparent.rs

```diff
@@ -438,63 +444,176 @@ pub(crate) fn put_received_transparent_utxo<P: consensus::Parameters>(
+        Err(SqliteClientError::AddressNotRecognized(
+            *output.recipient_address(),
+        ))
+    }
+}
+
+pub(crate) fn get_transparent_address_metadata<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
+    params: &P,
+    account_id: AccountId,
+    address: &TransparentAddress,
+) -> Result<Option<TransparentAddressMetadata>, SqliteClientError> {
+    let address_str = encode_transparent_address_p(params, address);
+
+    if let Some(di_vec) = conn
+        .query_row(
+            "SELECT diversifier_index_be FROM addresses
+             WHERE account_id = :account_id AND cached_transparent_receiver_address = :address",
+            named_params![":account_id": account_id.0, ":address": &address_str],
+            |row| row.get::<_, Vec<u8>>(0),
+        )
+        .optional()?
+    {
+        let address_index = address_index_from_diversifier_index_be(&di_vec)?;
+        let metadata = TransparentAddressMetadata::new(Scope::External.into(), address_index);
+        return Ok(Some(metadata));
+    }
+
+    if let Some((legacy_taddr, address_index)) =
+        get_legacy_transparent_address(params, conn, account_id)?
+    {
+        if &legacy_taddr == address {
+            let metadata = TransparentAddressMetadata::new(Scope::External.into(), address_index);
+            return Ok(Some(metadata));
+        }
+    }
+
+    // Search ephemeral addresses that have already been reserved.
+    if let Some(raw_index) = conn
+        .query_row(
+            "SELECT address_index FROM ephemeral_addresses
+             WHERE account_id = :account_id AND address = :address",
+            named_params![":account_id": account_id.0, ":address": &address_str],
+            |row| row.get::<_, u32>(0),
+        )
+        .optional()?
+    {
+        let address_index = NonHardenedChildIndex::from_index(raw_index).unwrap();
+        return Ok(Some(ephemeral::metadata(address_index)));
+    }
+
+    // We intentionally don't check for unreserved ephemeral addresses within the gap
+    // limit here. It's unnecessary to look up metadata for addresses from which we
+    // can spend.
+    Ok(None)
+}
+
+/// Attempts to determine the account that received the given transparent output.
+///
+/// The following three locations in the wallet's key tree are searched:
+/// - Transparent receivers that have been generated as part of a Unified Address.
+/// - Transparent ephemeral addresses that have been reserved or are within
+///   the gap limit from the last reserved address.
+/// - "Legacy transparent addresses" (at BIP 44 address index 0 within an account).
+///
+/// Returns `Ok(None)` if the transparent output's recipient address is not in any of the
+/// above locations. This means the wallet considers the output "not interesting".
+pub(crate) fn find_account_for_transparent_output<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
+    params: &P,
+    output: &WalletTransparentOutput,
```

Comment:
Why does this take `WalletTransparentOutput` instead of just `TransparentAddress`? It doesn't appear to need any of the other parts of the output.

Reply:
Also should `TransparentAddressMetadata` include the account id? That way one could just use `get_transparent_address_metadata`.

Reply:
> Why does this take `WalletTransparentOutput` instead of just `TransparentAddress`? It doesn't appear to need any of the other parts of the output.

Yes that is simpler, I will make this change.

> Also should `TransparentAddressMetadata` include the account id? That way one could just use `get_transparent_address_metadata`.

`TransparentAddressMetadata` possibly should include the account id, but it doesn't help much here because `get_transparent_address_metadata` takes the account id as an argument (and relies on having it), while here we don't know the account id. It could be possible to merge them but I'll leave it as-is for now.

Reply:
Changed.